data_structures/linked_list.ts

```diff
@@ -0,0 +1,475 @@
+   * ```
+   */
+  pop(): T | undefined {
+    if (!this.#tail) {
+      return undefined;
+    }
+
+    const node = this.#tail;
+    this.#tail = node.prev;
+
+    if (this.#tail) {
+      this.#tail.next = null;
+    } else {
+      this.#head = null;
+    }
+
+    --this.#length;
+    return node.value;
+  }
+
+  /**
+   * Appends a new value to the front of the linked list.
+   *
+   * The complexity of this operation is O(1).
+   *
+   * @param value The value to be added.
+   *
+   * @example Usage
+   * ```ts
+   * import { LinkedList } from "@std/data-structures";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const list = new LinkedList();
+   * assertEquals(list.length, 0);
+   *
+   * list.unshift(1);
+   * list.unshift(2);
+   * list.unshift(3);
+   *
+   * assertEquals(list.length, 3);
+   * assertEquals([...list], [3, 2, 1]);
+   * ```
+   */
+  unshift(value: T): void {
+    const node: LinkedNode<T> = { value, prev: null, next: this.#head };
+
+    if (this.#head) {
+      this.#head.prev = node;
+    } else {
+      this.#tail = node;
+    }
+
+    this.#head = node;
+    ++this.#length;
+  }
+
+  /**
+   * Removes the first node of the linked list and returns its value, if any.
+   *
+   * The complexity of this operation is O(1).
+   *
+   * @returns The removed node's value, `undefined` if this list is empty.
+   *
+   * @example Usage
+   * ```ts
+   * import { LinkedList } from "@std/data-structures";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const list = LinkedList.from(["a", "b", "c", "d"]);
+   * assertEquals(list.length, 4);
+   *
+   * assertEquals(list.shift(), "a");
+   * assertEquals(list.shift(), "b");
+   * assertEquals(list.shift(), "c");
+   *
+   * assertEquals(list.length, 1);
+   * ```
+   */
+  shift(): T | undefined {
+    if (!this.#head) {
+      return undefined;
+    }
+
+    const node = this.#head;
+    this.#head = node.next;
+
+    if (this.#head) {
+      this.#head.prev = null;
+    } else {
+      this.#tail = null;
+    }
+
+    --this.#length;
+    return node.value;
+  }
+
+  /**
+   * Inserts the given value to the specified index.
+   *
+   * The complexity of this operation is linear O(n).
+   *
+   * @param value The value to insert.
+   * @param index The position to insert.
+   * @throws If `index < 0` or `index > length`
+   *
+   * @example Usage
+   * ```ts
+   * import { LinkedList } from "@std/data-structures";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const list = LinkedList.from([2, 4, 6, 8]);
+   * assertEquals([...list], [2, 4, 6, 8]);
+   *
+   * list.insert(1, 0);
+   * assertEquals([...list], [1, 2, 4, 6, 8]);
+   *
+   * list.insert(3, 2);
+   * assertEquals([...list], [1, 2, 3, 4, 6, 8]);
+   *
+   * list.insert(5, 4);
+   * assertEquals([...list], [1, 2, 3, 4, 5, 6, 8]);
+   *
+   * list.insert(7, 6);
+   * assertEquals([...list], [1, 2, 3, 4, 5, 6, 7, 8]);
+   * ```
+   */
+  insert(value: T, index: number): void {
```

Comment:
If we want to follow the array convention this probably should be combined with `remove` to `splice`